### DISCUSSION TITLE
Use of Poetry to enhance F´ installation

### DISCUSSION THREAD

**Discussion by ThibFrgsGmz** [nasa/fprime#1005] [Show and tell]
Hi there 👋,

I have deployed on my [F´ fork](https://github.com/ThibFrgsGmz/fprime/tree/feat/poetry-mngt) the **Poetry** [tool](https://python-poetry.org/) which presents an elegant way to manage virtual environments and dependencies for development projects.

I think it is interesting to set it up on the project because it can make the installation of F' even easier, reducing the number of commands to write that would be reduced to : ``poetry install && peotry shell``.

Here is a demonstration of the installation:

## Installation of Peotry

Peotry is available via Pip but it is advisable to use the supplied installer which will isolate it from the rest of the system. Under GNU/Linux or macOS systems you should run the following command:

```bash
curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python -
```

See [here](https://python-poetry.org/docs/#installation) for other OS.

## Project configuration

Since we are instantiating Poetry on an existing project, in this case F´ , we will configure it with the command ``poetry init``:

```bash
$ cd fprime
$ poetry init
```

This results in an interactive configuration of the project.
We declare information such as the project name, license and dependencies.

```cmake
$ poetry init

This command will guide you through creating your pyproject.toml config.

Package name [fprime]:
Version [0.1.0]: 2.0.0
Description []:  F' - A flight software and embedded systems framework
Author [ThibFrgsGmz <xxx.yyy@zzz.com>, n to skip]:  Michael Starch
License []:  Apache 2.0 License
Compatible Python versions [^3.8]:

Would you like to define your main dependencies interactively? (yes/no) [yes]
You can specify a package in the following forms:
  - A single name (requests)
  - A name and a constraint (requests@^2.23.0)
  - A git url (git+https://github.com/python-poetry/poetry.git)
  - A git url with a revision (git+https://github.com/python-poetry/poetry.git#develop)
  - A file path (../my-package/my-package.whl)
  - A directory (../my-package/)
  - A url (https://example.com/packages/my-package-0.1.0.tar.gz)

Search for package to add (or leave blank to continue): fprime-tools
Found 20 packages matching fprime-tools

Enter package # to add, or the complete package name if it is not listed:
 [0] fprime-tools
 [1] fprime
 [2] fprime-gds
 [3] comment-out-code
 [4] ndmath
 [5] tools
 [6] gdprox
 [7] guoya-tools
 [8] bts_tools
 [9] time-tools
 > 0
Enter the version constraint to require (or leave blank to use the latest version):
Using version ^2.0.2 for fprime-tools

Add a package: fprime-gds
Found 20 packages matching fprime-gds

Enter package # to add, or the complete package name if it is not listed:
 [0] fprime-gds
 [1] fprime
 [2] gds
 [3] fprime-tools
 [4] comment-out-code
 [5] gds-metrics
 [6] gds3xtrude
 [7] py2gds
 [8] gds-tools
 [9] ndmath
 > 0
Enter the version constraint to require (or leave blank to use the latest version):
Using version ^2.0 for fprime-gds

Add a package:

Would you like to define your development dependencies interactively? (yes/no) [yes] no
Generated file

[tool.poetry]
name = "fprime"
version = "2.0.0
description = "F' - A flight software and embedded systems framework"
authors = ["Michael Starch"]
license = "Apache 2.0 License"

[tool.poetry.dependencies]
python = "^3.8"
fprime-tools = "^2.0.2"
fprime-gds = "^2.0"

[tool.poetry.dev-dependencies]

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"


Do you confirm generation? (yes/no) [yes]
```

At the end of this configuration no restructuring of the project is done, we get a pyprojet.toml file allowing to manage it with Peotry:

```
$ cat pyproject.toml
[tool.poetry]
name = "fprime
version = "2.0.0"
description = "F' - A flight software and embedded systems framework"
authors = ["Michael Starch"]
license = "Apache 2.0 License"

[tool.poetry.dependencies]
python = "^3.8"
fprime-tools = "^2.0.2"
fprime-gds = "^2.0"

[tool.poetry.dev-dependencies]

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```

This is the file that should be committed on F´ to manage it with Peotry.

**Note:** at the installation of the project by Peotry we will also note the creation of a locking file **poetry.lock**. This file contains the list of installed modules with their version number and would ensure that all developers use the same versions of the modules. The commit of this file is valuable because it will ensure that all contributors who configure the project use exactly the same versions of the dependencies that are officially used.

## New installation procedure for F´ with Peotry

With the pyprojet.toml file the installation procedure of F´ is simplified.

The developer will only have to run the ``poetry install`` command to install F´ . This installation goes beyond a simple installation of the modules listed in the pyproject.toml file, this command will also create the virtual environment in which we will be able to develop.

```
poetry install
Creating virtualenv fprime-mS5qxmBr-py3.8 in /home/ThibFrgsGmz/.cache/pypoetry/virtualenvs
Updating dependencies
Resolving dependencies... (2.9s)

Writing lock file

Package operations: 41 installs, 0 updates, 0 removals

```

We can see that the virtual environment has been created and we can access its information with the ``poetry env info`` command:

```
$ poetry env info

Virtualenv
Python: 3.8.5
Implementation: CPython
Path: /home/ThibFrgsGmz/.cache/pypoetry/virtualenvs/fprime-mS5qxmBr-py3.8
Valid: True

System
Platform: linux
OS: posix
Python: /usr
```

To access this virtual environment we just have to type the command ``peotry shell``:

```
$ poetry shell
Spawning shell within /home/ThibFrgsGmz/.cache/pypoetry/virtualenvs/fprime-mS5qxmBr-py3.8
ThibFrgsGmz@Thib:/mnt/c/Users/Thibault/peotry-test/fprime$ . /home/ThibFrgsGmz/.cache/pypoetry/virtualenvs/fprime-mS5qxmBr-py3.8/bin/activate
```

So the installation of F´ would be done with two commands: ``poetry install && peotry shell``.
    
If in the future we want to add a new production or development dependency we have the following commands:
- for production ``poetry add packageX``
- for development ``poetry add -D packageY``.

By analogy for removal we have the commands ``peotry remove packageX`` and ``peotry remove -D packageY``.

---

What do you think at first sight of this implementation? Do you think that it would add value to the project?

In my opinion this would probably improve the user experience, but it goes far beyond that.

---

**Comment by capsulecorplab** [comment]
Link to [F´ fork](https://github.com/ThibFrgsGmz/fprime/tree/feat/poetry-mngt) is broken

**Reply by ThibFrgsGmz** [reply]
Ah I think I deleted the branch a while ago since no one seemed interested.
I'll see next weekend if I can do it again. From what I remember, it was super simple to do.
